Add tests

johnhunter opened the pull request:
This PR adds the [Jest](https://jestjs.io/) test framework and some tests to cover the plugin interface to Eleventy.

The wish for tests was mentioned in the readme, and with recent regressions due to breaking changes in Eleventy #5 I felt it would be useful include automation against the plugin interface. I've raised this as a draft as it involved some refactoring of existing code to make it more testable.

## What was done

### Add Jest testing framework

Initially I spiked using [ava](https://www.npmjs.com/package/ava), but that would have involved additional dependencies for mocking and module mocking. Additionally, the mocks required running tests serially which isn't ava's default mode. Eventually I settled on the bigger but more complete Jest (which I'm more familiar with).

The following conventions followed:

- Tests colocated with code (e.g. `index.js`, `index.test.js`). Some prefer a separate test directory, but IMO its clearer when tests are missing, and require paths are simpler.
- Added a lint npm script, which is run with the test task.  This ensures lint and tests pass before a publish.
- Generate a coverage report on `npm test`. This is useful to see where more tests might be needed.
- All test and lint related files are excluded from the npm package

### Refactor to extract testable modules

I've broken apart the `index.js` - some functions moved to `utils`, and the class extracted to its own file. No functional change but it makes it possible to test each piece in isolation.

### Add tests against the Eleventy interface

I've added tests for everything that touches Eleventy. The Eleventy interface is mocked - and that's defined separately so we can mock different versions. I haven't added tests for the `EleventyLoad`, no reason not to but I don't think we need to if we just want to test compatibility with Eleventy.

These tests do not cover eleventy@1.0.0 yet. I have a PR #6 to address and whichever of these PRs is merged first I'll rebase and update the pending PR to cover the 1.0.0 interface.

Hope you find this useful @gregives - I'm happy to address issues or requested changes.

johnhunter commented:
@gregives best reviewed by commit 🙂

johnhunter commented on EleventyLoad.js:
```diff
@@ -0,0 +1,159 @@
+const path = require("path");
+const { emitFile, DEBUG_STRING } = require("./utils");
+
+module.exports = class EleventyLoad {
```
Extracted as-is from index.js

gregives commented:
@johnhunter I love this, thank you so much for your contribution! I've reviewed commit-by-commit, it all looks good to me. If you rebase this pull request to include your other changes, then we'll get this merged 😄

Would you like to be a collaborator of eleventy-load? I don't have much time to work on eleventy-load anymore but from the quality of your pull requests I'd be happy to make you a collaborator and give you publish permission as well!

johnhunter commented:
Thanks @gregives I'll rebase this to master and add testing for @1.0.0 at some point, might take me week or so. Feedback welcome in the meantime.

johnhunter commented:
> @johnhunter I love this, thank you so much for your contribution! I've reviewed commit-by-commit, it all looks good to me. If you rebase this pull request to include your other changes, then we'll get this merged 😄
> 
> Would you like to be a collaborator of eleventy-load? I don't have much time to work on eleventy-load anymore but from the quality of your pull requests I'd be happy to make you a collaborator and give you publish permission as well!

Thanks for the prompt response. I'm happy to be a contributor 😀  - although I might not have a lot of time to devote. I really liked the approach of leveraging the Webpack rules approach. I'm not using Eleventy in commercial work so It's more a free time thing at the moment but happy to help out.

gregives commented:
> Thanks for the prompt response. I'm happy to be a contributor 😀  - although I might not have a lot of time to devote. I really liked the approach of leveraging the Webpack rules approach. I'm not using Eleventy in commercial work so It's more a free time thing at the moment but happy to help out.

I've added you as a contributor on GitHub, please don't feel pressured to put any more time into it though — it will just reduce the effort for you to make a change!

johnhunter commented:
Rebased to master to pick up the v1xx fix #6 

I've improved the 11ty mocking - `this` is different for transform and shortcode callbacks. Also added the mocking for v1xx. The tests run for each version.

johnhunter commented:
@gregives hope you don't mind reviewing - its good to get a 2nd set of eyes before merging.

gregives commented on README.md:
```diff
@@ -89,6 +89,16 @@ Loaders can process any file, from a text file to images. See a list of loaders
 
 I'd love some help adding tests to eleventy-load and growing the ecosystem of loaders. If you'd like to contribute, get in touch with me!
 
+### Development
+
+- Clone this repo and run `npm install`
+- Configure your IDE to run prettier/eslint on save.
+- Run `npm run test:watch` which will run tests on changed files while you work.
+- Please add tests for any new features. Run `npm test` to lint, run tests and get a coverage report.
+
+It's a pain to manually test changes against a running 11ty instance, especially for different versions. The tests make this simpler by mocking the interface to 11ty for each major version. This gives reasonable confidence that the plugin is compatible with each major 11ty release.
```
Love this documentation 💯

gregives commented on index.js:
```diff
@@ -238,6 +45,8 @@ module.exports = function (config, options) {
 
   // Clear cache on re-runs
+  // TODO: update deprecated event name v0.11.0
+  // https://www.11ty.dev/docs/events/#eleventy.beforewatch
```
Do we need to change this to watch the new `eleventy.beforeWatch` as well as the old `beforeWatch`?

johnhunter replied:
Might be worth it, although its deprecated rather than removed. I suspect it will go in v3. I've left it for now - I thought I might look at options for detecting the 11ty version the plugin is running against.

gregives reviewed (approved):
Really happy with this pull request and the tests it introduces, thanks!
